core/client/fs/src/main/java/alluxio/client/file/FileInStream.java

```diff
@@ -521,402 +185,92 @@ private int positionedReadInternal(long pos, byte[] b, int off, int len) throws
-    boolean firstSeekOutsideFirstBlock =
-        mPositionState.getPos() == 0 && mCurrentBlockInStream == null && !isInCurrentBlock;
-    if (!firstSeekOutsideFirstBlock && canCacheToLocalWorker()) {
-      // Make sure that mCurrentBlockInStream and mCurrentCacheStream is updated.
-      // mPos is not updated here.
-      updateStreams();
 
-      // Cache till pos if seeking forward within the current block. Otherwise cache the whole
-      // block.
-      if (isInCurrentBlock && pos > mPositionState.getPos()) {
-        readCurrentBlockToPos(pos);
-      } else {
-        readCurrentBlockToEnd();
-      }
-
-      // Early return if we are at pos already. This happens if we seek forward with caching
-      // enabled for this block.
-      if (mPositionState.getPos() == pos) {
-        return;
+    if (mBlockInStream != null && mBlockInStream.remaining() == 0) { // current stream is done
+      closeBlockInStream();
+    }
+
+    /* Create a new stream to read from mPosition. */
+    // Calculate block id.
+    long blockId = mStatus.getBlockIds().get(Math.toIntExact(mPosition / mBlockSize));
+    // Create stream
+    mBlockInStream = mBlockStore.getInStream(mOptions.getBlockInfo(blockId), mOptions);
+    // Set the stream to the correct position.
+    long offset = mPosition % mBlockSize;
+    mBlockInStream.seek(offset);
+  }
+
+  private void closeBlockInStream() throws IOException {
+    if (mBlockInStream != null) {
+      // Clean up the current stream.
+      WorkerNetAddress dataSource = mBlockInStream.getAddress();
+      long blockId = mBlockInStream.getId();
+      mBlockInStream.close();
+      mBlockInStream = null;
+
+      // Send an async cache request to a worker based on read type and passive cache options.
+      boolean cache = mOptions.getOptions().getReadType().isCache();
+      boolean passiveCache = Configuration.getBoolean(PropertyKey.USER_FILE_PASSIVE_CACHE_ENABLED);
+      long channelTimeout = Configuration.getMs(PropertyKey.USER_NETWORK_NETTY_TIMEOUT_MS);
+      if (cache) {
+        // Construct the async cache request
+        Protocol.AsyncCacheRequest request =
+            Protocol.AsyncCacheRequest.newBuilder().setBlockId(blockId)
+                .setOpenUfsBlockOptions(mOptions.getOpenUfsBlockOptions(blockId))
+                .setSourceHost(dataSource.getHost()).setSourcePort(dataSource.getDataPort())
+                .build();
+        WorkerNetAddress worker;
+        if (passiveCache && mContext.hasLocalWorker()) { // send request to local worker
+          worker = mContext.getLocalWorker();
+        } else { // send request to data source
+          worker = dataSource;
+        }
+        Channel channel = mContext.acquireNettyChannel(worker);
+        try {
+          NettyRPCContext rpcContext =
+              NettyRPCContext.defaults().setChannel(channel).setTimeout(channelTimeout);
+          NettyRPC.call(rpcContext, new ProtoMessage(request));
```

Comment:
any way to see a warning message if this netty RPC fails?

Reply:
The method is blocking so it would throw an exception. I change it to log warn if any part of the async cache fails since it is best effort.